Fix version formatting to exclude .0 from EmulationVersion

ballista01 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind bug

#### What this PR does / why we need it:
This PR fixes the formatting of version strings displayed in leader-election leases:

- `EmulationVersion` is now displayed strictly as `major.minor`, omitting any trailing `.0` patches, prerelease, or build metadata.
- `BinaryVersion` is explicitly formatted as `major.minor.patch`, always including the patch version but omitting prerelease and build metadata (Same as before but removed dependency on semver methods).

To ensure clarity and consistency:

- Removed unnecessary intermediate semver parsing steps (`semver.ParseTolerant`) for EmulationVersion.
- Introduced two helper methods on the version type:
  - `MajorMinor()`: returns a version containing only the major and minor components.
  - `MajorMinorPatch()`: returns a version with major, minor, and patch.
- Both methods reset metadata and are safe to print via `.String()`.


#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #130933

#### Special notes for your reviewer:
- Maintained compatibility with existing logic and avoided changing internal version representation.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required.
-->
```release-note
Fixes version formatting for EmulationVersion displayed in leader-election leases.
```

ballista01 committed:
simplify EmulationVersion handling by avoiding using semver.ParseTolerant

siyuanfoundation commented:
Thanks @ballista01 for contributing! 
`effectiveVersion.String()` is not really invoked anywhere. 
For the leader election lease, the format is actually determined by the `semver.FinalizeVersion()` method here https://github.com/kubernetes/kubernetes/blob/ef838fcc3fa3b43c2e67b4e43d71b6720fc61898/cmd/kube-scheduler/app/server.go#L234. 

/cc @Jefftree

seans3 commented:
/triage accepted

ballista01 commented:
Hi, I’d like to clarify a few things before finalizing the patch.

I originally modified `effectiveVersion.String()` and added a `majorMinor()` helper with tests, but later realized that method isn’t used for EmulationVersion output. Let me know if you’d prefer I remove that code since it’s no longer necessary.

While reviewing the code more closely, I noticed there’s also a function-style `FinalizeVersion(string)`. It calls `Version.String()`, and still outputs the patch version. I tried removing the patch portion globally by modifying `String()`, but that broke etcd-migrate tests. Although I can fix that, I’m concerned it may cause other potential compatibility issues.

So instead, I’ve updated both the method and function versions of `FinalizeVersion` to drop the patch when it’s zero, without touching `Version.String()`. Since these seem to be user-facing display helpers, I believe this achieves the intended formatting change with minimal risk.

Just wanted to check if this approach aligns with what you were expecting.

Thanks!

siyuanfoundation commented:
@Jefftree 
Why do the versions in leaderelection need to be converted into semver first, then string? Can we just use `c.ComponentGlobalsRegistry.EffectiveVersionFor(basecompatibility.DefaultKubeComponent).EmulationVersion().String()` directly? That would drop the patch if it is 0.

ballista01 committed:
add methods for formatting EmulationVersion and BinaryVersion

ballista01 commented:
/re-test

ballista01 commented:
/retest

jpbetz commented on staging/src/k8s.io/apimachinery/pkg/util/version/version.go:
```diff
@@ -333,6 +333,34 @@ func (v *Version) String() string {
+// MajorMinorString returns the version as "major.minor",
+// omitting patch, pre-release, and build metadata.
+// Used for EmulationVersion display in leases and logs.
+func (v *Version) MajorMinorString() string {
```
We already have `String()` receiver that will print the string out correctly if provided the correct components.  Let's leverage this.  Maybe allow for `MajorMinor().String()` by adding a `MajorMinor() *Version` receiver? Same for `MajorMinorPatch`.

ballista01 replied:
Yeah, reusing the logic in `String()` seems more maintainable.

ballista01 committed:
add MajorMinor and MajorMinorPatch methods for version formatting

ballista01 commented on staging/src/k8s.io/component-base/compatibility/version.go:
```diff
@@ -95,7 +95,7 @@ func (m *effectiveVersion) String() string {
 	}
 	return fmt.Sprintf("{BinaryVersion: %s, EmulationVersion: %s, MinCompatibilityVersion: %s}",
-		m.BinaryVersion().String(), m.EmulationVersion().String(), m.MinCompatibilityVersion().String())
+		m.BinaryVersion().String(), majorMinor(m.EmulationVersion()).String(), majorMinor(m.MinCompatibilityVersion()).String())
```
Removed but I kept the unit test `TestEffectiveVersionString` which reflects the current behavior (EmulationVesion and MinCompatibilityVersion are MajorMinor formats). Is it good?

siyuanfoundation commented on cmd/kube-controller-manager/app/controllermanager.go:
```diff
@@ -293,23 +292,17 @@ func Run(ctx context.Context, c *config.CompletedConfig) error {
-		if err != nil {
-			return err
-		}
+		binaryVersionStr := c.ComponentGlobalsRegistry.EffectiveVersionFor(basecompatibility.DefaultKubeComponent).BinaryVersion().MajorMinorPatch().String()
```
@Jefftree is it intended to drop release info for binary version?

ballista01 replied:
https://github.com/ballista01/kubernetes/blob/fe7af1c68bc0c772e78d57b5c38ac4ae8ef122ab/vendor/github.com/blang/semver/v4/semver.go#L64-L66

It's the original behavior but if we decide to include the build & release info, the current `String()` method is all we need.

Jefftree replied:
The build and release info is not used for CLE's semver comparison so I think it's fine.

siyuanfoundation replied:
In that case, you can also just use `WithPreRelease("")`. No need to add a new method of `MajorMinorPatch` here

siyuanfoundation commented on cmd/kube-controller-manager/app/controllermanager.go:
```diff
@@ -293,23 +292,17 @@ func Run(ctx context.Context, c *config.CompletedConfig) error {
-			return err
-		}
+		binaryVersionStr := c.ComponentGlobalsRegistry.EffectiveVersionFor(basecompatibility.DefaultKubeComponent).BinaryVersion().MajorMinorPatch().String()
+		emulationVersionStr := c.ComponentGlobalsRegistry.EffectiveVersionFor(basecompatibility.DefaultKubeComponent).EmulationVersion().MajorMinor().String()
```
`EmulationVersion` only contains `MajorMinor`, so `MajorMinor()` is not necessary. 
https://github.com/kubernetes/kubernetes/blob/e3b9541b235e47e7afd7edef403e1187f1e62a16/staging/src/k8s.io/component-base/compatibility/version.go#L109

jpbetz commented:
/approve

Deferring to @siyuanfoundation for LGTM

ballista01 committed:
Removed unnecessary MajorMinor method

ballista01 commented:
/retest

ballista01 committed:
Merge branch 'kubernetes:master' into fix/emulation-version-format
